dulwich/tests/test_porcelain.py

```diff
@@ -1577,10 +1577,230 @@ def test_resetfile_with_dir(self):
+
+        porcelain.commit(
+            repo=self.repo.path,
+            message=b"init",
+            author=b"author <email>",
+            committer=b"committer <email>",
+        )
+
+        # Setup target repo cloned from temp test repo
+        clone_path = tempfile.mkdtemp()
+        self.addCleanup(shutil.rmtree, clone_path)
+        target_repo = porcelain.clone(
+            self.repo.path, target=clone_path, errstream=errstream
+        )
+        try:
+            self.assertEqual(target_repo[b"HEAD"], self.repo[b"HEAD"])
+        finally:
+            target_repo.close()
+
+        # create a second file to be pushed back to origin
+        handle, fullpath = tempfile.mkstemp(dir=clone_path)
+        os.close(handle)
+        porcelain.add(repo=clone_path, paths=[fullpath])
+        porcelain.commit(
+            repo=clone_path,
+            message=b"push",
+            author=b"author <email>",
+            committer=b"committer <email>",
+        )
+
+        # Setup a non-checked out branch in the remote
+        refs_path = b"refs/heads/foo"
+        new_id = self.repo[b"HEAD"].id
+        self.assertNotEqual(new_id, ZERO_SHA)
+        self.repo.refs[refs_path] = new_id
+
+        # Push to the remote
+        porcelain.push(
+            clone_path,
+            "origin",
+            b"HEAD:" + refs_path,
+            outstream=outstream,
+            errstream=errstream,
+        )
+
+        self.assertEqual(
+            target_repo.refs[b"refs/remotes/origin/foo"],
+            target_repo.refs[b"HEAD"],
+        )
+
+        porcelain.checkout(target_repo, b"origin/foo")
+        original_id = target_repo[b"HEAD"].id
+
+        expected_refs = {
+            b"HEAD": original_id,
+            b"refs/heads/master": original_id,
+            b"refs/heads/foo": original_id,
+            b"refs/remotes/origin/foo": original_id,
+            b"refs/remotes/origin/HEAD": new_id,
+            b"refs/remotes/origin/master": new_id,
+        }
+        self.assertEqual(expected_refs, target_repo.get_refs())
+
+        return target_repo
+
+    def test_checkout_remote_branch(self):
+        self._checkout_remote_branch()
+
+    def test_checkout_remote_branch_then_master_then_remote_branch_again(self):
+        target_repo = self._checkout_remote_branch()
+        self.assertEqual(b"foo", porcelain.active_branch(target_repo))
+        _commit_file_with_content(target_repo, 'bar', 'something')
+        self.assertTrue(os.path.isfile(os.path.join(target_repo.path, 'bar')))
+
+        porcelain.checkout(target_repo, b"master")
+        self.assertEqual(b"master", porcelain.active_branch(target_repo))
+        self.assertFalse(os.path.isfile(os.path.join(target_repo.path, 'bar')))
+
+        porcelain.checkout(target_repo, b"origin/foo")
+        self.assertEqual(b"foo", porcelain.active_branch(target_repo))
+        self.assertTrue(os.path.isfile(os.path.join(target_repo.path, 'bar')))
+
+
+def tree(root_dir):
```

Comment:
This isn't actually used?

Reply:
Whoops, used it for debugging.

Reply:
[90180da](https://github.com/jelmer/dulwich/pull/1148/commits/90180dadacaf2fff71cb9373dd9e4614a9a1eb4f)